### PR TITLE
Refactor shared storage.

### LINKED ISSUE #86: Shared data and shared queue leak on VM change
looks like this repo is the right place to solve

ref: 
- https://github.com/envoyproxy/envoy/issues/13835
- https://github.com/istio/istio/issues/27783

### PR DESCRIPTION
**Pull request by mathetake**
refactor the shared storage code:

- separate out SharedQueue functionalities from ShareData class since they are independent of each other. It looks like they do not need to even share mutex.
- write unit tests on each class

In a sequel PR, I would tackle the leakage issue

### CODE REVIEW

**Review by PiotrSikora — COMMENTED** [review]
Could you avoid mixing big refactoring / moving code with bug fixes? While I don't mind the refactoring, it makes it impossible to review the fix.

**Comment by mathetake** [comment]
right. Thanks. I will narrow down this PR scope only to refactoring and adding tests

**Comment by mathetake** [comment]
@PiotrSikora updated the PR description and title, and made ready for review! PTAL

File: test/shared_storage_test.cc [hunk 1]
```diff
@@ -0,0 +1,80 @@
+// Copyright 2020 Google LLC
```
Comment (mathetake) [inline test/shared_storage_test.cc:1]: would be better to make tests cover concurrent access

**Comment by PiotrSikora** [comment]
Thanks! Since, as you noted, they are completely independent, should we split them into `src/shared_data.*` and `src/shared_queue.*`? or do you prefer to leave them together?

**Comment by mathetake** [comment]
> should we split them into src/shared_data.* and src/shared_queue.*?

I will split them! Thanks for the comment

**Comment by PiotrSikora** [comment]
This is just split + tests, right? No code changes?

**Comment by mathetake** [comment]
YES, no code changes

**Review by PiotrSikora — APPROVED** [review]

**Comment by PiotrSikora** [comment]
Thanks!

**Pull request merged**